src/lib.rs

```diff
@@ -132,11 +137,22 @@ where
 ///     B(EventB),
 /// }
 ///
-/// impl EventData for DomainEvents {}
+/// impl EventData for DomainEvents {
+///   fn event_type(&self) -> String {
+///     match &self {
+///         DomainEvents::A(_) => "namespace.A",
+///         DomainEvents::B(_) => "namespace.B",
+///         _ => panic!("Woops"),
+///     }.into()
+///   }
+/// }
 ///
 /// fn main() {}
 /// ```
-pub trait EventData: Serialize + DeserializeOwned {}
+pub trait EventData: Serialize + DeserializeOwned {
+    /// Must return the type of an event
+    fn event_type(&self) -> String;
+}
```

Comment:
On `TestEvents`, move the impl of `event_type` here using the Serde trick I pasted in another comment. That way it can live in the default impl and we don't have to write as much code.